### PR TITLE
Non-determinism fix: ClassLoaders are now transformed in the Model Checking mode

### LINKED ISSUE #412: ClassLoaders are not transformed in ModelChecking mode
ClassLoaders are not transformed in ModelChecking mode. That leads to non-determinism and potential deadlocks.

### PR DESCRIPTION
**Pull request by avpotapov00**
closes #412, #377, #405

### CODE REVIEW

**Review by ndkoval — CHANGES_REQUESTED** [review]
Hi, @avpotapov00, thanks for the fix!

Could you please add a simple test that reveals the bug and measure the performance impact of this change? Intuitively, when running a single test, re-instrumenting all classes with `ClassLoader` in their names can increase the running time by several seconds.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/BufferedChannelTest.kt [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import kotlin.coroutines.cancellation.CancellationException
+
+/**
+ * Checks that [bug with not transformed ClassLoader-s](https://github.com/JetBrains/lincheck/issues/412) is resolved.
+ */
+@Param.Params(
+    Param(name = "elem", gen = IntGen::class, conf = "1:3"),
+    Param(name = "closeToken", gen = IntGen::class, conf = "1:3")
+)
+class BufferedChannelTest {
```
[the remark this answers is not in the document]
Comment (avpotapov00) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/BufferedChannelTest.kt:32]: It's not a good decision to put the implementation into the current file, it becomes huge.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt [hunk 2]
```diff
@@ -351,7 +356,8 @@ internal object LincheckClassFileTransformer : ClassFileTransformer {
         // once they are used in the testing code.
         if (!INSTRUMENT_ALL_CLASSES &&
             instrumentationMode == MODEL_CHECKING &&
-            internalClassName.canonicalClassName !in instrumentedClasses) {
+            internalClassName.canonicalClassName !in instrumentedClasses &&
+            !isClassLoader(internalClassName)) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt:360]: Please fix the indentation

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/TransformationUtils.kt [hunk 3]
```diff
@@ -347,3 +347,5 @@ internal const val ASM_API = Opcodes.ASM9
 internal val STRING_TYPE = getType(String::class.java)
 internal val CLASS_TYPE = getType(Class::class.java)
 internal val CLASS_FOR_NAME_METHOD = Method("forName", CLASS_TYPE, arrayOf(STRING_TYPE))
+
+internal fun isClassLoader(className: String): Boolean = className.contains("ClassLoader")
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/TransformationUtils.kt:351]: Please rename the function to indicate a possibility of false positives (or make it `isNotClassLoader` and add the comment about false positives where you use the function)

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/BufferedChannelTest.kt [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import org.jetbrains.kotlinx.lincheck.paramgen.IntGen
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingOptions
+import org.junit.Test
+import kotlin.coroutines.cancellation.CancellationException
+
+/**
+ * Checks that [bug with not transformed ClassLoader-s](https://github.com/JetBrains/lincheck/issues/412) is resolved.
+ */
+@Param.Params(
+    Param(name = "elem", gen = IntGen::class, conf = "1:3"),
```
[the remark this answers is not in the document]
Comment (ndkoval) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/BufferedChannelTest.kt:29]: Do you need these parameters?

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/BufferedChannelTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndkoval) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/BufferedChannelTest.kt:32]: It's huge anyway, but this is just a copy-and-paste from somewhere (which should be indicated)

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt [hunk 5]
```diff
@@ -351,7 +357,8 @@ internal object LincheckClassFileTransformer : ClassFileTransformer {
         // once they are used in the testing code.
         if (!INSTRUMENT_ALL_CLASSES &&
             instrumentationMode == MODEL_CHECKING &&
-            internalClassName.canonicalClassName !in instrumentedClasses) {
+            internalClassName.canonicalClassName !in instrumentedClasses &&
+            !containsClassloaderInName(internalClassName)) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt:361]: Should it be `!containsClassloaderInName` or `containsClassloaderInName` here?

Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt:361]: Please add a comment that we always transform class loaders

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/IncorrectBufferedChannel.kt [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+
+package org.jetbrains.kotlinx.lincheck_test.transformation.channel
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.CancellableContinuation
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+/**
+ * Incorrect implementation of the BufferedChannel, intentionally provided by Daria Shutina.
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/channel/IncorrectBufferedChannel.kt:19]: Please add a reference to the issue, and put the code and the test in the same file.

**Review by ndkoval — APPROVED** [review]

**Pull request merged**